### PR TITLE
Simplify the cache decoding graph

### PR DESCRIPTION
**Pull request by mattdangerw**
This updates our `CachedMultiHeadAttention` layer to avoid slicing the input to a dynamic length during generative decoding. Instead, our cache is always a fixed length, as are the keys and values after the cache is applied.

Overall this ends up being both a nice simplification and speedup. After patching in a fix for #779, generating 25 sequences of length 256 with the base model goes from 52s -> 33s.

### CODE REVIEW

File: keras_nlp/samplers/sampler.py [hunk 1]
```diff
@@ -172,18 +172,6 @@ def _validate_prompt_and_mask(self, prompt, mask):
         prompt = tf.ragged.boolean_mask(prompt, mask)
         return prompt, mask
 
-    def _validate_token_probability_fn(
```
Comment (mattdangerw) [inline keras_nlp/samplers/sampler.py:175]: I think we need to ditch this. It just isn't going to be performant enough. Especially right now outside of the compiled function.

File: keras_nlp/layers/transformer_decoder.py [hunk 2]
```diff
@@ -272,7 +271,18 @@ def call(
         x = decoder_sequence  # Intermediate result.
 
         # Compute self attention mask.
-        self_attention_mask = compute_causal_mask(decoder_sequence)
+        batch_size = tf.shape(decoder_sequence)[0]
+        input_length = output_length = tf.shape(decoder_sequence)[1]
+        # When doing cached decoding, our attention mask is as long as the cache
```
Comment (chenmoneygithub) [inline keras_nlp/layers/transformer_decoder.py:276]: This looks a bit confusing to me - the input seq should be of len 1, so the attention mask is of size `[batch_size, seq_len (cache_len), 1]`. When we say "as long as the cache", it took me a while to understand what it means...

Also we may want to comment on the shape of `attention_mask`, I thought it should be `[batch_size, 1, seq_len (cache_len)]`, but running the code I found it's `[batch_size, seq_len (cache_len), 1]`.

File: keras_nlp/layers/transformer_decoder_test.py [hunk 3]
```diff
@@ -266,56 +266,49 @@ def test_mask_propagation_without_cross_attention(self):
...
-        cache = tf.zeros(
-            [batch_size, 2, seq_len, num_heads, hidden_dim // num_heads]
-        )
-        # Build the intial cache.
-        initial_seq_len = 2
-        initial_inputs = x[:, :initial_seq_len, :]
+        x = tf.random.uniform(shape=[batch_size, seq_len, num_heads * head_dim])
+        cache = tf.zeros([batch_size, 2, seq_len, num_heads, head_dim])
         outputs = tf.zeros_like(x)
-        output, cache = layer(
```
Comment (chenmoneygithub) [inline keras_nlp/layers/transformer_decoder_test.py:287]: [optional] Although we have unified the first call with other calls, shall we keep the testing code? Like for the first call we slice until some position, like `x[:5]`. My intention is to mock the actual situation, where the first call has `seq_len > 1`, with following calls having `seq_len == 1`

File: keras_nlp/layers/transformer_layer_utils.py [hunk 4]
```diff
@@ -18,21 +18,11 @@
...
-    mask = tf.reshape(mask, (1, input_shape[1], input_shape[1]))
-    mult = tf.concat(
-        [
-            tf.expand_dims(batch_size, -1),
-            tf.constant([1, 1], dtype=tf.int32),
-        ],
-        axis=0,
-    )
-    return tf.tile(mask, mult)
+def compute_causal_mask(batch_size, input_length, output_length, cache_index=0):
```
Comment (chenmoneygithub) [inline keras_nlp/layers/transformer_layer_utils.py:21]: We may want to add docstring to this function now. `input_length` and `output_length` is a bit confusing, like for the decoder, the input sequence is of length `1`, but it is the `output_length` in this function.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 5]
```diff
@@ -356,28 +289,36 @@ def generate(
...
+
+        # Tokenize.
+        prompt = self.preprocessor.tokenizer(prompt)
+
+        # Create and seed the cache before generation.
+        token_ids = prompt
+        if prompt.shape.rank == 1:
+            token_ids = tf.RaggedTensor.from_tensor(prompt[tf.newaxis, :])
+        token_ids = token_ids.to_tensor(shape=(None, max_length))
+        padding_mask = tf.ones_like(token_ids, dtype=tf.bool)
```
Comment (chenmoneygithub) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:309]: This is a dense tensor with all 1s. The functionality is unaffected because causal mask is used, but it might hurt our readability. Shall set the correct value of `padding_mask`, or add a comment here saying "the value of `padding_mask` does not matter because causal mask is applied"?

**Review by chenmoneygithub — CHANGES_REQUESTED** [review]
Thanks Matt! It works pretty well! Dropped a few comments on the code itself.

File: keras_nlp/layers/transformer_decoder.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/layers/transformer_decoder.py:276]: Yeah, I should probably clearer here. I will reword to make clear the attention mask maps from a source sequence of length `cache_length` to a dest sequence of length `1`.

File: keras_nlp/layers/transformer_layer_utils.py (hunk 4 above)
Comment (mattdangerw) [reply keras_nlp/layers/transformer_layer_utils.py:21]: sg!

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 5 above)
Comment (mattdangerw) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:309]: I'll add the comment for now.

In future cleanup I would actually like the remove the mask updates we are doing in the sampler method, as they are totally unnecessary and just slow us down! Essentially the only masks we need to care about during generation are a causal mask and the mask of where we the original inputs were in our sequence tensor. Slowly scattering ones into our padding mask is just busywork.

I do struggle with how to make this clear to users though.

File: keras_nlp/layers/transformer_decoder_test.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/layers/transformer_decoder_test.py:287]: I think the logic of testing that you could pass an entire "chunk" to seed a cache is encapsulated well in the "no loop" call we compare with. I would focus on brevity while actually testing the dimension we want here.

**Comment by mattdangerw** [comment]
Thanks for the review! Addressed all comments.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 6]
```diff
@@ -356,28 +289,38 @@ def generate(
...
+        token_ids = prompt
+        if prompt.shape.rank == 1:
+            token_ids = tf.RaggedTensor.from_tensor(prompt[tf.newaxis, :])
+        token_ids = token_ids.to_tensor(shape=(None, max_length))
+        # Pass a padding mask of all ones when seeing the cache. The mask will
+        # not affect cached key/values for input tokens we care about.
+        padding_mask = tf.ones_like(token_ids, dtype=tf.bool)
+        batch_size = tf.shape(token_ids)[0]
+        cache = self.build_empty_cache(batch_size, max_length)
+        _, cache = self.call_with_cache(token_ids, padding_mask, cache, 0)
```
Comment (chenmoneygithub) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:314]: I am thinking we can fold this initial call into the `sampler` call. When `cache_index=0`, the code knows there is no cached computed before. But let's discuss it in your one pager and launch this PR!

**Review by chenmoneygithub — APPROVED** [review]
Thanks Matt!

File: keras_nlp/layers/transformer_decoder_test.py [hunk 7]
```diff
@@ -266,56 +266,49 @@ def test_mask_propagation_without_cross_attention(self):
...
-        graph_cached_outputs = graph_call(initial_seq_len, cache, outputs)
-        normal_outputs = layer(decoder_sequence=x)
-        self.assertAllClose(cached_outputs, normal_outputs)
-        self.assertAllClose(graph_cached_outputs, normal_outputs)
+            return outputs, cache
+
+        call = call if eager else tf.function(call)
+        output, cache = call(outputs, cache)
+
+        no_loop_outputs, no_loop_cache = layer(x, cache=cache)
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/layers/transformer_decoder_test.py:309]: Yeah, good call! We can test both. Updated.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 6 above)
Comment (mattdangerw) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:314]: I think we should move compilation up the model for a whole host of reasons--not insignificantly including consistency with core Keras in terms of how it handles compilation. But agreed let's keep discussing this as a follow up!
